core/math/triangle_mesh.cpp

```diff
@@ -510,6 +510,222 @@ bool TriangleMesh::intersect_ray(const Vector3 &p_begin, const Vector3 &p_dir, V
 	return inters;
 }
 
+bool TriangleMesh::intersect_convex_shape(const Plane *p_planes, int p_plane_count) const {
+	uint32_t *stack = (uint32_t *)alloca(sizeof(int) * max_depth);
+
+	//p_fully_inside = true;
+
+	enum {
+		TEST_AABB_BIT = 0,
+		VISIT_LEFT_BIT = 1,
+		VISIT_RIGHT_BIT = 2,
+		VISIT_DONE_BIT = 3,
+		VISITED_BIT_SHIFT = 29,
+		NODE_IDX_MASK = (1 << VISITED_BIT_SHIFT) - 1,
+		VISITED_BIT_MASK = ~NODE_IDX_MASK,
+
+	};
+
+	int level = 0;
+
+	PoolVector<Triangle>::Read trianglesr = triangles.read();
+	PoolVector<Vector3>::Read verticesr = vertices.read();
+	PoolVector<BVH>::Read bvhr = bvh.read();
+
+	const Triangle *triangleptr = trianglesr.ptr();
+	const Vector3 *vertexptr = verticesr.ptr();
+	int pos = bvh.size() - 1;
+	const BVH *bvhptr = bvhr.ptr();
+
+	stack[0] = pos;
+	while (true) {
+
+		uint32_t node = stack[level] & NODE_IDX_MASK;
+		const BVH &b = bvhptr[node];
+		bool done = false;
+
+		switch (stack[level] >> VISITED_BIT_SHIFT) {
+			case TEST_AABB_BIT: {
+
+				bool valid = b.aabb.intersects_convex_shape(p_planes, p_plane_count);
+				if (!valid) {
```

Comment:
Just a style trick to avoid having too many indentation levels: here you could put `continue;` in the `if (!valid) { }` block, and then remove the `else {}` condition and keep its content unindented (also finishing with `continue;` of course).

Like this:
```c++
if (!valid) {
	stack[level] = (VISIT_DONE_BIT << VISITED_BIT_SHIFT) | node;
	continue;
}

if (b.face_index >= 0) {

	const Triangle &s = triangleptr[b.face_index];
...
} else {
	stack[level] = (VISIT_LEFT_BIT << VISITED_BIT_SHIFT) | node;
}
continue;
```